core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,221 @@
+
+  private val MAX_FAILURES_PER_EXEC = conf.get(config.MAX_FAILURES_PER_EXEC)
+  private val MAX_FAILED_EXEC_PER_NODE = conf.get(config.MAX_FAILED_EXEC_PER_NODE)
+  val EXECUTOR_RECOVERY_MILLIS = BlacklistTracker.getBlacklistExpiryTime(conf)
+
+  // a count of failed tasks for each executor.  Only counts failures after tasksets complete
+  // successfully
+  private val executorIdToFailureCount: HashMap[String, Int] = HashMap()
+  private val executorIdToBlacklistExpiryTime: HashMap[String, Long] = new HashMap()
+  private val nodeIdToBlacklistExpiryTime: HashMap[String, Long] = new HashMap()
+  private val _nodeBlacklist: AtomicReference[Set[String]] = new AtomicReference(Set())
+  private var nextExpiryTime: Long = Long.MaxValue
+
+  def start(): Unit = {}
+
+  def stop(): Unit = {}
+
+  def expireExecutorsInBlacklist(): Unit = {
+    val now = clock.getTimeMillis()
+    // quickly check if we've got anything to expire from blacklist -- if not, avoid doing any work
+    if (now > nextExpiryTime) {
+      val execsToClear = executorIdToBlacklistExpiryTime.filter(_._2 < now).keys
+      if (execsToClear.nonEmpty) {
+        logInfo(s"Removing executors $execsToClear from blacklist during periodic recovery")
+        execsToClear.foreach { exec => executorIdToBlacklistExpiryTime.remove(exec) }
+      }
+      if (executorIdToBlacklistExpiryTime.nonEmpty) {
+        nextExpiryTime = executorIdToBlacklistExpiryTime.map{_._2}.min
+      } else {
+        nextExpiryTime = Long.MaxValue
+      }
+      val nodesToClear = nodeIdToBlacklistExpiryTime.filter(_._2 < now).keys
+      if (nodesToClear.nonEmpty) {
+        logInfo(s"Removing nodes $nodesToClear from blacklist during periodic recovery")
+        nodesToClear.foreach { node => nodeIdToBlacklistExpiryTime.remove(node) }
+        // make a copy of the blacklisted nodes so nodeBlacklist() is threadsafe
+        _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+      }
+    }
+ }
+
+  def taskSetSucceeded(
+      failuresByExec: HashMap[String, FailureStatus],
+      scheduler: TaskSchedulerImpl): Unit = {
+    // if any tasks failed, we count them towards the overall failure count for the executor at
+    // this point.
+    failuresByExec.foreach { case (exec, newFailures) =>
+      val prevFailures = executorIdToFailureCount.getOrElse(exec, 0)
+      val newTotal = prevFailures + newFailures.totalFailures
+
+      if (newTotal >= MAX_FAILURES_PER_EXEC) {
+        logInfo(s"Blacklisting executor $exec because it has $newTotal" +
+          s" task failures in successful task sets")
+        val now = clock.getTimeMillis()
+        val expiryTime = now + EXECUTOR_RECOVERY_MILLIS
+        executorIdToBlacklistExpiryTime.put(exec, expiryTime)
+        executorIdToFailureCount.remove(exec)
+        if (expiryTime < nextExpiryTime) {
+          nextExpiryTime = expiryTime
+        }
+
+        val node = scheduler.getHostForExecutor(exec)
```

Comment:
Note this results in the node not really being blacklisted in the next stage even though it might have other executors on it that should be blacklisted.